velox/vector/DecodedVector.cpp

```diff
@@ -155,14 +192,15 @@ void DecodedVector::combineWrappers(
   int32_t levelCounter = 0;
   for (;;) {
     if (numLevels != -1 && ++levelCounter == numLevels) {
-      baseVector_ = values;
+      baseVectorShared_ = values;
```

Comment:
Its a bit unclear what the contract here is between different decode APIs. Like, will only the API that takes a VectorPtr save a reference to the base?
If yes, then there are some instances like this one where a reference is always taken. This can then affect existing use cases that currently use the BaseVector API and do not expect the base to outlive the original vector. One such case that is causing test failures is in the ExprTest suite. This can be fixed by clearing the decodedVector by calling reset() before adding it to the decodedVector cache. https://github.com/facebookincubator/velox/blob/ccb5094211888fbbd0633dad2a25db864e1c6c3d/velox/expression/EvalCtx.h#L797-L801

Reply:
I am also thinking about avoid holding ownership of the smart pointer, but the cast from specific `shared_ptr` (e.g. `FlatVectorPtr`) to `VectorPtr` is annoying.  Maybe I will just keep the ownership exclusively through the new method, and avoid holding onto it in any old calls.